services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -251,6 +257,168 @@ public URL getUrl(GetUrlRequest getUrlRequest) {
+
+        if ("s3".equalsIgnoreCase(uri.getScheme())) {
+            return parseAwsCliStyleUri(uri);
+        }
+
+        return parseStandardUri(uri);
+    }
+
+    private S3Uri parseStandardUri(URI uri) {
+        String bucket = null;
+        String key = null;
+        String region = null;
+        boolean isPathStyle = false;
+        Map<String, List<String>> queryParams = new HashMap<>();
+        String path = uri.getPath();
+
+        if (uri.getHost() == null) {
+            throw new IllegalArgumentException("Invalid S3 URI: no hostname: " + uri);
+        }
+
+        Matcher matcher = ENDPOINT_PATTERN.matcher(uri.getHost());
+        if (!matcher.find()) {
+            throw new IllegalArgumentException("Invalid S3 URI: hostname does not appear to be a valid S3 endpoint: " + uri);
+        }
+
+        String prefix = matcher.group(1);
+        if (prefix == null || prefix.isEmpty()) {
+            isPathStyle = true;
+
+            if (!path.isEmpty() && !"/".equals(path)) {
+                int index = path.indexOf('/', 1);
+
+                if (index == -1) {
+                    // No trailing slash, e.g., "https://s3.amazonaws.com/bucket"
+                    bucket = path.substring(1);
+                } else {
+                    bucket = path.substring(1, index);
+                    if (index != path.length() - 1) {
+                        key = path.substring(index + 1);
+                    }
+                }
+            }
+        } else {
+            bucket = prefix.substring(0, prefix.length() - 1);
+            if (path != null && !path.isEmpty() && !"/".equals(path)) {
+                key = path.substring(1);
+            }
+        }
+
+        if (!"amazonaws".equals(matcher.group(2))) {
+            region = matcher.group(2);
+        }
+
+        String queryPart = uri.getQuery();
+        if (queryPart != null) {
+            parseQuery(queryParams, queryPart);
+        }
+
+        Region uriRegion = region != null ? Region.of(region) : null;
+
+        return S3Uri.builder()
+                    .uri(uri)
+                    .bucket(bucket)
+                    .key(key)
+                    .region(uriRegion)
+                    .isPathStyle(isPathStyle)
+                    .queryParams(queryParams)
+                    .build();
+    }
+
+    private S3Uri parseAwsCliStyleUri(URI uri) {
+        String key = null;
+        String bucket = uri.getAuthority();
+        Region region = null;
+        boolean isPathStyle = false;
+        Map<String, List<String>> queryParams = new HashMap<>();
+        String path = uri.getPath();
+
+        if (bucket == null) {
+            throw new IllegalArgumentException("Invalid S3 URI: bucket not included: " + uri);
+        }
+
+        if (path.length() > 1) {
+            key = path.substring(1);
+        }
+
+        String queryPart = uri.getQuery();
+        if (queryPart != null) {
+            parseQuery(queryParams, queryPart);
```

Comment:
I'm not sure if S3 URI supports query params. https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-bucket-intro.html#:~:text=access%20by%20HTTPS.-,Accessing%20a%20bucket%20using%20S3%3A//,-Some%20AWS%20services